### PR TITLE
Add initial support for traffic-policy feature

### LINKED ISSUE #578: Add traffic-policy feature in eos_cli_config_gen
**Issue Type**

- Role enhancement

**Is your feature request related to a problem? Please describe.**

Allow configuration of traffic-policy on interface as per EOS documentation: https://eos.arista.com/eos-4-25-0f/support-for-traffic-policy-on-interfaces/

**Describe the solution you'd like**

Implement the following data model to generate EOS configuration

```yaml
traffic_policies:
  ipv4:
    BLUE-C1-POLICY:
      match:
        BLUE-C1-POLICY-01:
          source:
            prefixes:
              - 10.0.0.0/8
              - 192.168.0.0/16
          protocols:
            tcp:
              src_port: "1,10-20"
            icmp:
          actions:
            log: true
            traffic_class: 8
        BLUE-C1-POLICY-02:
          source:
            prefix_lists:
              - PL-BOGONS
          protocols:
            tcp:
              dst_port: "80,443"
            icmp:
          actions:
            count: MyCounter
            dscp: 60
        BLUE-C1-POLICY-02:
          source:
            prefix_lists:
              - PL-BOGONS
          protocols:
            tcp:
          actions:
            count: DROP-COUNTER
            drop: true
```

### PR DESCRIPTION
**Pull request by titom73**
## Change Summary

Add initial support for new traffic-policy in eos_cli_config_gen

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

Fixes #578

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes

__Supported features:__

- Support prefix and prefix-list as source
- Support protocol with source and destination port when applicable
- Support the following actions:
  - Count
  - Set traffic class
  - Set DSCP
  - Logging
  - Drop
- Generate documentation as well

__Know limitations:__

- Only IPv4
- Support source OR destination ports not a mix of both
- Does not set correct TCAM profile and requires to configure it by using `tcam_profile` knob

__Data model example__

```yaml
---
traffic_policies:
  options:
    counter_per_interface: < true | false >
  field_sets:
    ipv4:
      < PREFIX FIELD SET NAME >:
        - < IPv4 prefix 01>
        - < IPv4 prefix 02>
        - < IPv4 prefix 03>
    ipv6:
      < PREFIX FIELD SET NAME >:
        - < IPv6 prefix 01>
        - < IPv6 prefix 02>
        - < IPv6 prefix 03>
    ports:
      < L4 PORT FIELD SET NAME >: "< vlan range >"
  policies:
    < TRAFFIC POLICY NAME >:
      matchs:
        < TRAFFIC POLICY ITEM >:
          type: < ipv4 | ipv6 >
          source:
            prefixes:
              - < prefix 01 >
              - < prefix 02 >
            prefix_lists:
              - < Field Set List 01 >
              - < Field Set List 02 >
          ttl: "< ttl range>"
          # The 'fragment' command is not supported when 'source port'
          # or 'destination port' command is configured
          fragment:
            offset: "< fragment offset range >"
          protocols:
            tcp:
              src_port: "< port range >"
              dst_port: "< port range >"
              src_field: "< L4 port range field set >"
              dst_field: "< L4 port range field set >"
              flags:
                - established
                - initial
            icmp:
              icmp_type:
                - < ICMP message type >
                - < ICMP message type >
            udp:
              src_port: "< port range >"
              dst_port: "< port range >"
              src_field: "< L4 port range field set >"
              dst_field: "< L4 port range field set >"
            ahp:
            bgp:
            icmp:
            igmp:
            ospf:
            pim:
            rsvp:
            vrrp:
            # The 'protocol neighbors' subcommand is not supported when any
            # other match subcommands are configured
            neighbors:
          actions:
            dscp: < dscp code value >
            traffic_class: < traffic class id >
            count: < counter name >
            drop: < true | false (default false) >
            # Only supported when action is set to drop
            log: < true | false (default false) >
          # Last resort policy
          default_actions:
            < ipv4 | ipv6 >:
              dscp: < dscp code value >
              traffic_class: < traffic class id >
              count: < counter name >
              drop: < true | false (default false) >
              # Only supported when action is set to drop
              log: < true | false (default false) >
```

Report overview [available](https://github.com/titom73/ansible-avd/blob/features/traffic-policies/ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/traffic-policies.md#traffic-policies-1) 

## How to test

Refer to molecule eos_cli_config_gen scenario

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://www.avd.sh/docs/contributing/) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly

### CODE REVIEW

**Comment by titom73** [comment]
Todo list:

- [x] Add neighbors support for `protocol bgp`
- [x] Add ICMP code for `protocol icmp`
- [x] IP destination field
- [x] TTL / Fragment / DSCP as condition
- [x] Port Field set for source and destination
- [x] Prefix Field-set
- [x] TCP Flags

**Comment by titom73** [comment]
@carlbuchmann , @aphillipps : Ready for Data model review and testing

Interfaces configuration to apply a traffic-policy will be added later when available on vEOS-LAB

**Comment by titom73** [comment]
__TODO__

- Create data model for interface to attach traffic-policy
- Require PR #623 to avoid conflicts since ethernet and PO have been refactor

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/vlan-interfaces.md
Comment (carlbuchmann) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/vlan-interfaces.md]: can you re-run molecule. It looks like some artifacts need to be updated!

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/hosts
Comment (carlbuchmann) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/hosts]: remove text likely from resolving merge conflict.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/traffic-policies.j2
Comment (carlbuchmann) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/traffic-policies.j2]: missing newline EOF.

**Review by carlbuchmann — CHANGES_REQUESTED** [review]
please see comments! The data model looks good!

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/traffic-policies.j2
Comment (titom73) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/traffic-policies.j2]: Resolved with b335ad8

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/hosts
Comment (titom73) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/hosts]: Merge conflict resolved with 3235a29

**Review by carlbuchmann — APPROVED** [review]
LGTM - but would also like a second review before merging!

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/traffic-policies.md
Comment (ankudinov) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/traffic-policies.md:24]: there is only one L3 header under L2 header "Management". Can we remove one?
It's cosmetic though.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/traffic-policies.md
Comment (ankudinov) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/traffic-policies.md:52]: ideally header should be removed from the document if the section is empty. Cosmetic.

**Comment by titom73** [comment]
Documentation is not generated due to devel changes. Working on it

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/traffic-policies.yml
Comment (ankudinov) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/traffic-policies.yml]: matches?

Comment (ankudinov) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/traffic-policies.yml]: or match_list?

**Review by ankudinov — COMMENTED** [review]
LGTM, except few cosmetic things (see comments).

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/traffic-policies.md
Comment (titom73) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/documentation/devices/traffic-policies.md:24]: Should be part of another PR related to documentation build as it is not related to traffic policy?

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/traffic-policies.yml
Comment (carlbuchmann) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/traffic-policies.yml]: what about `match_items:`, since it's a dictionary?

Comment (titom73) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/traffic-policies.yml]: I would vote for the most concise approach as data model for this feature is already big.

So I vote for matches or matches whatever is the best for you

**Review by carlbuchmann — APPROVED** [review]
LGTM!

**Pull request merged**
